# Machine Learning Environment for the Isabelle Proof Assistant

<div style="max-width: 1000px;">

## Demostration of usage
To demostrate the ability of small-step and big-step proof correctness verification in training and evaluating of Prover, we select five questions from miniF2F, raw data formated in the form of LEGO-prover data.
### Questions
|problem name | informal statement | formal statement |
| ----------- | ------------------ | ---------------- |
| mathd_numbertheory_3 | What is the units digit of the sum of the squares of the first nine positive integers? Show that it is 5. | ```theorem mathd_numbertheory_3 :\n  \"(\<Sum> x < 10. ((x + 1)^2)) mod 10 = (5::nat)\"``` |
| mathd_algebra_419 | What is the value of $-a-b^2+3ab$ if $a=-1$ and $b=5$? Show that it is -39. | ```theorem mathd_algebra_419:\n  fixes a b :: real\n  assumes \"a = -1\"\n and \"b = 5\"\n  shows \"-a - b^2 + 3 * (a * b) = -39\"``` |
| numbertheory_3pow2pownm1mod2pownp3eq2pownp2 | Show that for positive integer $n$, the remainder of $3^{2^n} - 1 \equiv 2^{n+2} \mod 2^{n+3}$. | ```theorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes \"0 < n\" \n  shows \"(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)\"``` |
| aime_1983_p1 | Let $x$, $y$ and $z$ all exceed $1$ and let $w$ be a positive number such that $\\log_x w = 24$, $\\log_y w = 40$ and $\\log_{xyz} w = 12$. Find $\\log_z w$. Show that it is 060. | ```theorem aime_1983_p1:\n  fixes x y z w :: nat\n  assumes ht : \"1 < x \\<and> 1 < y \\<and> 1 < z\"\n    and hw : \"0 \\<le> w\"\n    and h0 : \"ln w / ln x = 24\"\n    and h1 : \"ln w / ln y = 40\"\n    and h2 : \"ln w / ln (x * y * z) = 12\"\n  shows \"ln w / ln z = 60\"``` |
| amc12b_2020_p22 | What is the maximum value of $\frac{(2^t-3t)t}{4^t}$ for real values of $t?$ $\textbf{(A)} \frac{1}{16} \qquad\textbf{(B)} \frac{1}{15} \qquad\textbf{(C)} \frac{1}{12} \qquad\textbf{(D)} \frac{1}{10} \qquad\textbf{(E)} \frac{1}{9}$ Show that it is $\textbf{(C)} \frac{1}{12}$. | ```theorem amc12b_2020_p22:\n  fixes t :: real\n  shows \"((2 powr t - 3 * t) * t) / (4 powr t) \<le> 1 / 12\"``` |
</div>

In [1]:
# Raw data
questions = [
    {
        "problem_name": "mathd_numbertheory_3",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "What is the units digit of the sum of the squares of the first nine positive integers? Show that it is 5.",
        "informal_proof": "We can reduce the amount of work we have to do in this problem by realizing that the units digit of the sum of the squares is the units digit of the sum of the units digits of the squares. In other words, the units digit of $1^2+2^2+\\ldots+9^2$ is the units digit of $1+4+9+6+5+6+9+4+1=45$, which is $5$.",
        "formal_statement": "theorem mathd_numbertheory_3 :\n  \"(\\<Sum> x < 10. ((x + 1)^2)) mod 10 = (5::nat)\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_numbertheory_3 imports Complex_Main\nbegin\n\ntheorem mathd_numbertheory_3 :\n  \"(\\<Sum> x \\<le> 9. (x^2)) mod 10 = (5::nat)\"\n  by eval\n\nend "
    },
    {
        "problem_name": "mathd_algebra_419",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "What is the value of $-a-b^2+3ab$ if $a=-1$ and $b=5$? Show that it is -39.",
        "informal_proof": "Plugging in the given values yields $-a-b^2+3ab=-(-1)-5^2+3(-1)(5)=1-25-15=-39$.",
        "formal_statement": "theorem mathd_algebra_419:\n  fixes a b :: real\n  assumes \"a = -1\"\n    and \"b = 5\"\n  shows \"-a - b^2 + 3 * (a * b) = -39\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_algebra_419 imports Complex_Main\nbegin\n\ntheorem mathd_algebra_419:\n  fixes a b :: real\n  assumes \"a = -1\"\n    and \"b = 5\"\n  shows \"-a - b^2 + 3 * (a * b) = -39\"\n  using assms by auto\n\nend"
    },
    {
        "problem_name": "numbertheory_3pow2pownm1mod2pownp3eq2pownp2",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "Show that for positive integer $n$, the remainder of $3^{2^n} - 1 \\equiv 2^{n+2} \\mod 2^{n+3}$.",
        "informal_proof": "We can prove this by induction on n. For the base case, $n=1$, when the result obviously holds.\nFor the inductive case, the hypothesis is $3^{2^n} - 1 \\equiv 2^{n+2} \\mod 2^{n+3}$, so we can obtain a natural number $p$ where $3^{2^n} - 1 = 2^{n+2} + p * 2^{n+3} = 2^{n+2}(1+2p)$.\nThen we have $3^{2^{n+1}} = 3^{2^n*2} = (3^{2^n})^2 = (2^{n+2} * (1+2p) + 1)^2$.\nHence $3^{2^{n+1}} \\equiv (2^{n+2} * (1+2p))^2 + 2 * 2^{n+2} * (1+2p) + 1 \\equiv 2^{n+3}*(1+2p) + 1 \\equiv 1\\mod 2^{n+3}$.\nTherefore $3^{2^{n+1}} -1 \\equiv 0 \\mod 2^{n+3}$. By induction, the statement holds true.",
        "formal_statement": "theorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes \"0 < n\" \n  shows \"(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory numbertheory_3pow2pownm1mod2pownp3eq2pownp2 \n  imports Complex_Main \"HOL-Computational_Algebra.Computational_Algebra\"\n  \"HOL-Number_Theory.Number_Theory\"\nbegin\n\ntheorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes \"0 < n\" \n  shows \"(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)\"\n  using assms\nproof (induct n)\n  case (Suc n)\n  have ?case when \"n=0\" \n    using that by simp\n  moreover have ?case when \"n>0\"\n  proof -\n    define m::nat where \"m = 2^n\"\n    have \"(3 ^ 2 ^ n - 1) mod 2 ^ (n + 3) = (2::nat)^ (n + 2)\"\n      using Suc(1) that by simp\n    then have \"(3^m - 1) mod (8*m) = 4*m\"\n      unfolding m_def by (auto simp:algebra_simps power_add) \n    then obtain k where k0:\"3^m - 1 = (8*m)*k + 4*m\" \n      by (metis mult_div_mod_eq)\n    define M::nat where \"M = 4*m*k^2+m+4*m*k+k\"\n\n    have k:\"3^m = (8*m)*k + 4*m +1\" \n    proof -\n      have \"3^m\\<ge>(1::nat)\" by auto\n      then show ?thesis\n        using k0 by linarith\n    qed\n    have \"3 ^ 2 ^ Suc n - 1 = ((3::nat) ^ (m*2)) - 1\"\n      unfolding m_def by (auto simp:algebra_simps)\n    also have \"... = (3 ^ m)\\<^sup>2 - 1\"\n      unfolding power_mult by simp\n    also have \"... = ((8*m)*k + 4*m +1)^2 -1 \"\n      unfolding k by simp\n    also have \"... = (16*m)*M + 8*m\"\n      unfolding M_def by (auto simp:algebra_simps power2_eq_square)\n    finally have \"3 ^ 2 ^ Suc n - 1 = (16*m)*M + 8*m\" .\n    moreover have \"((16*m)*M + 8*m) mod (16*m) = 8*m\"\n      by simp\n    then have \"((16*m)*M + 8*m) mod 2 ^ (Suc n + 3) = 2 ^ (Suc n + 2)\"\n      unfolding m_def by (auto simp:algebra_simps power_add)\n    ultimately show ?thesis by auto\n  qed\n  ultimately show ?case by blast\nqed simp\n\nend   "
    },
    {
        "problem_name": "aime_1983_p1",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "Let $x$, $y$ and $z$ all exceed $1$ and let $w$ be a positive number such that $\\log_x w = 24$, $\\log_y w = 40$ and $\\log_{xyz} w = 12$. Find $\\log_z w$. Show that it is 060.",
        "informal_proof": "The [[logarithm]]ic notation doesn't tell us much, so we'll first convert everything to the equivalent exponential forms.\n\n$x^{24}=w$, $y^{40}=w$, and $(xyz)^{12}=w$. If we now convert everything to a power of $120$, it will be easy to isolate $z$ and $w$.\n\n$x^{120}=w^5$, $y^{120}=w^3$, and $(xyz)^{120}=w^{10}$.\n\nWith some substitution, we get $w^5w^3z^{120}=w^{10}$ and $\\log_zw=060$.",
        "formal_statement": "theorem aime_1983_p1:\n  fixes x y z w :: nat\n  assumes ht : \"1 < x \\<and> 1 < y \\<and> 1 < z\"\n    and hw : \"0 \\<le> w\"\n    and h0 : \"ln w / ln x = 24\"\n    and h1 : \"ln w / ln y = 40\"\n    and h2 : \"ln w / ln (x * y * z) = 12\"\n  shows \"ln w / ln z = 60\"",
        "formal_code": "(*\n  Authors: Wenda Li\n*)\n\ntheory aime_1983_p1 imports Complex_Main\nbegin\n\ntheorem aime_1983_p1:\n  fixes x y z w :: nat\n  assumes ht : \"1 < x \\<and> 1 < y \\<and> 1 < z\"\n    and hw : \"0 \\<le> w\"\n    and h0 : \"ln w / ln x = 24\"\n    and h1 : \"ln w / ln y = 40\"\n    and h2 : \"ln w / ln (x * y * z) = 12\"\n  shows \"ln w / ln z = 60\"\nproof -\n  define xx yy zz ww where \"xx=ln x\" and \"yy = ln y\" \n      and \"zz = ln z\" and \"ww = ln w\"\n  have \"xx= ww/24\" using h0 ht unfolding xx_def ww_def \n    by (auto simp:field_simps)\n  moreover have \"yy = ww / 40\" using ht h1 unfolding yy_def ww_def\n    by (auto simp:field_simps)\n  moreover have \"xx+yy+zz > 0\" \n    unfolding xx_def yy_def zz_def using ht \n    by (metis ln_gt_zero of_nat_1 of_nat_less_iff pos_add_strict)\n  then have \"ww = 12*(xx+yy+zz)\"using ht h2\n    unfolding xx_def yy_def zz_def ww_def\n    by (auto simp:field_simps ln_mult)\n  ultimately have \"ww = 12*(ww/24 + ww/40 + zz)\"\n    by blast\n  then have \"ww=60*zz\"\n    by (auto simp:field_simps)\n  then show ?thesis unfolding ww_def zz_def using ht by auto\nqed\n\nend\n"
    },
    {
        "problem_name": "amc12b_2020_p22",
        "category": "unknown",
        "metadata": {},
        "informal_statement": "What is the maximum value of $\\frac{(2^t-3t)t}{4^t}$ for real values of $t?$\n\n$\\textbf{(A)}\\ \\frac{1}{16} \\qquad\\textbf{(B)}\\ \\frac{1}{15} \\qquad\\textbf{(C)}\\ \\frac{1}{12} \\qquad\\textbf{(D)}\\ \\frac{1}{10} \\qquad\\textbf{(E)}\\ \\frac{1}{9}$ Show that it is \\textbf{(C)} \\frac{1}{12}.",
        "informal_proof": "We proceed by using AM-GM. We get $\\frac{(2^t-3t) + 3t}{2}$ $\\ge \\sqrt{(2^t-3t)(3t)}$. Thus, squaring gives us that $4^{t-1} \\ge (2^t-3t)(3t)$. Rembering what we want to find, we divide both sides of the inequality by the positive amount of $\\frac{1}{3\\cdot4^t}$. We get the maximal values as $\\frac{1}{12}$, and we are done.",
        "formal_statement": "theorem amc12b_2020_p22:\n  fixes t :: real\n  shows \"((2 powr t - 3 * t) * t) / (4 powr t) \\<le> 1 / 12\"",
        "formal_code": "theory amc12b_2020_p22\n  imports Complex_Main \"HOL-Library.Code_Target_Numeral\"\nbegin\n\ntheorem amc12b_2020_p22: fixes t :: real shows \"((2 powr t - 3 * t) * t) / (4 powr t) \<le> 1 / 12\" proof - have h1: \"4 * ((2 powr t - 3 * t) * (3 * t)) \<le> (2 powr t)^2\" by (smt (verit) power2_sum ring_class.ring_distribs(2) sum_squares_bound) have hpow: \"(2 powr t)^2 = 4 powr t\" by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) have h2: \"12 * ((2 powr t - 3 * t) * t) \<le> 4 powr t\" using h1 hpow by simp have pos: \"0 < 12 * (4 powr t)\" by simp have eq: \"((2 powr t - 3 * t) * t) / (4 powr t) = (12 * ((2 powr t - 3 * t) * t)) / (12 * (4 powr t))\" by (simp add: field_simps) have \"(12 * ((2 powr t - 3 * t) * t)) / (12 * (4 powr t)) \<le> (4 powr t) / (12 * (4 powr t))\" using h2 pos by simp also have \"... = 1 / 12\" by simp finally show ?thesis using eq by simp qed end"
    }
]

In [2]:
# Path setup: make the repo importable regardless of where Jupyter was launched.
# (Also requires the deps: pip install -r requirement.txt — at minimum httpx —
#  and the server running: docker compose exec isabelle-gym python -m server.app.main)
import sys, os

REPO_ROOT = None
_here = os.getcwd()
for _candidate in [_here, os.path.dirname(_here), os.path.dirname(os.path.dirname(_here))]:
    if os.path.exists(os.path.join(_candidate, "client", "async_client.py")):
        REPO_ROOT = _candidate
        if _candidate not in sys.path:
            sys.path.insert(0, _candidate)
        print("repo root on sys.path:", _candidate)
        break
if REPO_ROOT is None:
    raise RuntimeError(
        "IsabelleGym repo root not found — launch Jupyter from the repo (or demo/) directory"
    )


repo root on sys.path: /Users/winstonren/GitHub/IsabelleGym


In [3]:
from client.async_client import IsabelleGymAsyncClient
client = IsabelleGymAsyncClient()
proofs = [i.get("formal_code") for i in questions]
proofs

['(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_numbertheory_3 imports Complex_Main\nbegin\n\ntheorem mathd_numbertheory_3 :\n  "(\\<Sum> x \\<le> 9. (x^2)) mod 10 = (5::nat)"\n  by eval\n\nend ',
 '(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_algebra_419 imports Complex_Main\nbegin\n\ntheorem mathd_algebra_419:\n  fixes a b :: real\n  assumes "a = -1"\n    and "b = 5"\n  shows "-a - b^2 + 3 * (a * b) = -39"\n  using assms by auto\n\nend',
 '(*\n  Authors: Wenda Li\n*)\n\ntheory numbertheory_3pow2pownm1mod2pownp3eq2pownp2 \n  imports Complex_Main "HOL-Computational_Algebra.Computational_Algebra"\n  "HOL-Number_Theory.Number_Theory"\nbegin\n\ntheorem numbertheory_3pow2pownm1mod2pownp3eq2pownp2:\n  fixes n :: nat\n  assumes "0 < n" \n  shows "(3^(2^n) - 1) mod (2^(n + 3)) = (2::nat)^(n + 2)"\n  using assms\nproof (induct n)\n  case (Suc n)\n  have ?case when "n=0" \n    using that by simp\n  moreover have ?case when "n>0"\n  proof -\n    define m::nat where "m = 2^n"\n    have "(3 ^ 2 ^ n -

In [4]:
import sys
from pathlib import Path

ROOT = Path.cwd().parent
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

In [5]:
from evaluation.scripts.theory_splitter import (
    extract_theory_name,
    split_commands,
)

## Bigstep verification

In [14]:
result = await client.verify_bigstep_text(theory_name="mathd_numbertheory_3", theory_text=proofs[0])
result

<Response [200 OK]>

In [ ]:
for proof in proofs:
    result = await client.verify_bigstep_text(theory_name=extract_theory_name(proof), theory_text=proof)
    print(result)

## Smallstep verification

### Single proof verification
For this demostration, we will need to split the proof to stepwise tactics to simulate the real use case in prover training and evaluation. This step is by the same splitter used in the evaluation.

In [23]:
session_payload = await client.acquire_session(["Complex_Main"], "HOL")
session_payload

{'session_id': '55232a7b-6a62-4a66-a546-6c6d8c27d69f',
 'created_at': 1779913230.0128596,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': False,
 'lease_id': 'c8d82ff97d88'}

In [24]:
session_id = session_payload.get("session_id")
lease_id = session_payload.get("lease_id")
commands = split_commands(proofs[0])
commands

['(*\n  Authors: Wenda Li\n*)\n\ntheory mathd_numbertheory_3 imports Complex_Main\nbegin',
 'theorem mathd_numbertheory_3 :\n  "(\\<Sum> x \\<le> 9. (x^2)) mod 10 = (5::nat)"',
 'by eval',
 'end']

In [25]:
await client.enter_theory(session_id, "mathd_numbertheory_3", lease_id=lease_id)

{'success': True, 'message': 'Entered theory mathd_numbertheory_3'}

In [26]:
await client.execute_command(session_id, commands[0], lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.11754894256591797,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [27]:
await client.execute_command(session_id, commands[1], lease_id=lease_id)

{'success': True,
 'output': 'val it = (): unit',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. sum power2 {..9} mod 10 = 5'],
 'execution_time': 0.1082608699798584,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [28]:
await client.execute_command(session_id, "sledgehammer", lease_id=lease_id)

{'success': True,
 'output': 'val it = (): unit\nval it = (): unit\nSledgehammering...\nNo proof found',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. sum power2 {..9} mod 10 = 5'],
 'execution_time': 11.7899489402771,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [25]:
await client.execute_command(session_id, commands[2], lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.40366268157958984,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [26]:
await client.execute_command(session_id, commands[3], lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.10254931449890137,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

## Test sledgehammer

Earlier we invoked sledgehammer by sending the literal `sledgehammer` Isar command as a proof **step** (`execute_command(..., 'sledgehammer')`). Running it that way executes inside the document node and blocks the session's other ML channels (e.g. `open_subgoals`) until it finishes.

Here we use the **standalone sledgehammer channel** instead: `client.sledgehammer(session_id, ...)` hits the dedicated `/sledgehammer` endpoint. On the ML side it forks an Isabelle thread, so it does **not** run as a proof step, does not advance/modify the proof state (read-only), and does not block other sessions. It returns the proof-method suggestions (e.g. `by (metis ...)`) which the prover can then apply as the next step.

In [14]:
session_payload = await client.acquire_session(["Complex_Main"], "HOL")
session_payload

{'session_id': '034b2d04-94ab-425d-9cbb-fb7ae93b76f7',
 'created_at': 1780463041.5500674,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': True,
 'lease_id': '7690aba810aa'}

In [15]:
session_id = session_payload.get("session_id")
lease_id = session_payload.get("lease_id")

In [ ]:
# Test release session
await client.release_session(session_id, lease_id=lease_id)

{'success': True, 'session_id': '034b2d04-94ab-425d-9cbb-fb7ae93b76f7'}

In [16]:
await client.enter_theory(session_id, "amc12b_2020_p22", lease_id=lease_id)

{'success': True, 'message': 'Entered theory amc12b_2020_p22'}

In [17]:
await client.execute_command(session_id, 'theory amc12b_2020_p22 imports Complex_Main "HOL-Library.Code_Target_Numeral" begin', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.7222416400909424,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [18]:
await client.execute_command(session_id, 'theorem amc12b_2020_p22: fixes t :: real shows "((2 powr t - 3 * t) * t) / (4 powr t) ≤ 1 / 12"', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10210990905761719,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [19]:
await client.execute_command(session_id, 'proof -', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10212922096252441,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [20]:
await client.execute_command(session_id, 'have hpow: "(2 powr t)^2 = 4 powr t"', lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t)\\<^sup>2 = 4 powr t'],
 'execution_time': 0.10215520858764648,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [21]:
# Standalone sledgehammer channel: dedicated endpoint, not a proof step.
# Returns proof-method suggestions without modifying the proof state.
sh = await client.sledgehammer(session_id, timeout_s=60, lease_id=lease_id)
sh

{'success': True,
 'suggestions': ['metis found a proof...',
  'metis found a proof...',
  'metis: Try this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)',
  'metis: Duplicate proof',
  'zipperposition found a proof...',
  'zipperposition: Try this: by (metis numeral_Bit0_eq_double power2_eq_square powr_mult) (31 ms)',
  'Done',
  'Try this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)'],
 'raw_output': 'metis found a proof...\nmetis found a proof...\nmetis: Try this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)\nmetis: Duplicate proof\nzipperposition found a proof...\nzipperposition: Try this: by (metis numeral_Bit0_eq_double power2_eq_square powr_mult) (31 ms)\nDone\nTry this: by (metis powr_mult numeral_Bit0_eq_double power2_eq_square) (7 ms)',
 'execution_time': 21.489171504974365}

In [22]:
await client.execute_command(session_id, 'by (metis powr_mult numeral_Bit0_eq_double power2_eq_square)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have hpow: (2 powr t)⇧2 = 4 powr t',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.11284089088439941,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [23]:
await client.execute_command(session_id, 'have pos: "(0::real) < 4 powr t" by simp', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have pos: 0 < 4 powr t',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10291004180908203,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [39]:
await client.execute_command(session_id, 'have amgm: "2 * (2 powr t) * (6 * t) ≤ (2 powr t)^2 + (6 * t)^2"', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. 2 * 2 powr t * (6 * t) \\<le> (2 powr t)\\<^sup>2 + (6 * t)\\<^sup>2'],
 'execution_time': 0.10155200958251953,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [40]:
await client.execute_command(session_id, 'by (rule sum_squares_bound)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have amgm: 2 * 2 powr t * (6 * t) ≤ (2 powr t)⇧2 + (6 * t)⇧2',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10209131240844727,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [41]:
await client.execute_command(session_id, 'have rearr: "12 * ((2 powr t - 3 * t) * t) = 2 * (2 powr t) * (6 * t) - (6 * t)^2"', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. 12 * ((2 powr t - 3 * t) * t) = 2 * 2 powr t * (6 * t) - (6 * t)\\<^sup>2'],
 'execution_time': 0.10149717330932617,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [42]:
await client.execute_command(session_id, 'by (simp add: power2_eq_square algebra_simps)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have rearr:\n       12 * ((2 powr t - 3 * t) * t) = 2 * 2 powr t * (6 * t) - (6 * t)⇧2',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.1020505428314209,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [43]:
await client.execute_command(session_id, 'from amgm rearr hpow have key: "12 * ((2 powr t - 3 * t) * t) ≤ 4 powr t"', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. 12 * ((2 powr t - 3 * t) * t) \\<le> 4 powr t'],
 'execution_time': 0.10160684585571289,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [44]:
await client.execute_command(session_id, 'by linarith', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'have key: 12 * ((2 powr t - 3 * t) * t) ≤ 4 powr t',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10194897651672363,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [45]:
await client.execute_command(session_id, 'from key pos show ?thesis', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': ['1. (2 powr t - 3 * t) * t / 4 powr t \\<le> 1 / 12'],
 'execution_time': 0.10175180435180664,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [46]:
await client.execute_command(session_id, 'by (simp add: field_simps)', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'show (2 powr t - 3 * t) * t / 4 powr t ≤ 1 / 12\nSuccessful attempt to solve goal by exported rule:\n  (2 powr t - 3 * t) * t / 4 powr t ≤ 1 / 12',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.10207438468933105,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [47]:
await client.execute_command(session_id, 'qed', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': 'theorem amc12b_2020_p22:\n (2 powr ?t - 3 * ?t) * ?t / 4 powr ?t ≤ 1 / 12',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.1020059585571289,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [48]:
await client.execute_command(session_id, 'end', timeout=120.0, lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.10172724723815918,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

## Test session parallelism (`parallel_proofs` / `threads`)

Sessions are now created with configurable PIDE parallelism (`repl/src/main/scala/repl/session_manager.scala`):

- `ISABELLE_PARALLEL_PROOFS` (default `2`) — `2` forks nested `have`/`show` proof bodies so **independent `have`s are checked concurrently**; `1` = top-level proofs only; `0` = fully sequential.
- `ISABELLE_SESSION_THREADS` (default `4`) — caps the ML worker threads per session (deliberately not `0` = auto, so concurrent sessions don't oversubscribe cores and OOM the gateway).

Both are read from the **server's** environment at session creation — set them in `.env` at the repo root and restart the server to change them. Pooled sessions keep the options they were created with.

**The experiment:** one lemma with `N_HAVES` independent `have` steps, each proof sleeping `SLEEP_S` seconds inside an ML tactic before closing the goal. Submitted as **one chunk**, PIDE forks the `have` proofs, so wall time ≈ `ceil(N_HAVES / threads) * SLEEP_S`. The same steps submitted **one command at a time** form the sequential baseline (~`N_HAVES * SLEEP_S`), because each small-step waits for the node to finish checking before returning.

In [4]:
session_payload = await client.acquire_session(["Complex_Main"], "HOL")
session_id = session_payload.get("session_id")
lease_id = session_payload.get("lease_id")
session_payload

{'session_id': '9f22572d-ef71-4c95-af4d-95c157abe754',
 'created_at': 1781128631.4464405,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': False,
 'lease_id': '27b151213841'}

In [5]:
N_HAVES = 4
SLEEP_S = 2

def slow_have(i: int) -> str:
    # Independent goal; the ML tactic sleeps, then closes it with simp.
    # With parallel_proofs=2 these `by` bodies are forked and run concurrently.
    return (
        f'have h{i}: "{i} + 0 = ({i}::nat)"\n'
        f"  by (tactic ‹(OS.Process.sleep (Time.fromSeconds {SLEEP_S}); simp_tac @{{context}} 1)›)"
    )

parallel_chunk = "\n".join(
    ["lemma par_demo: True", "proof -"]
    + [slow_have(i) for i in range(1, N_HAVES + 1)]
    + ["show True by simp", "qed"]
)
print(parallel_chunk)

lemma par_demo: True
proof -
have h1: "1 + 0 = (1::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
have h2: "2 + 0 = (2::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
have h3: "3 + 0 = (3::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
have h4: "4 + 0 = (4::nat)"
  by (tactic ‹(OS.Process.sleep (Time.fromSeconds 2); simp_tac @{context} 1)›)
show True by simp
qed


In [6]:
await client.enter_theory(session_id, "Par_Demo", lease_id=lease_id)
await client.execute_command(session_id, "theory Par_Demo imports Complex_Main begin", lease_id=lease_id)

{'success': True,
 'output': '',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 0.3339200019836426,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [7]:
import time

t0 = time.perf_counter()
res = await client.execute_command(session_id, parallel_chunk, timeout=120.0, lease_id=lease_id)
parallel_wall = time.perf_counter() - t0
print(f"one chunk, {N_HAVES} parallel haves: {parallel_wall:.1f}s "
      f"(sequential would be ~{N_HAVES * SLEEP_S}s of sleep alone)")
res

one chunk, 4 parallel haves: 6.9s (sequential would be ~8s of sleep alone)


{'success': True,
 'output': 'have h1: 1 + 0 = 1\nhave h2: 2 + 0 = 2\nhave h3: 3 + 0 = 3\nhave h4: 4 + 0 = 4\nshow True\nSuccessful attempt to solve goal by exported rule:\n  True\ntheorem par_demo: True',
 'error': None,
 'subgoal_error': None,
 'subgoals': [],
 'execution_time': 6.663745641708374,
 'mode': None,
 'diagnostics': [],
 'failure_location': None,
 'theory_verified': False}

In [8]:
# Sequential baseline: same steps, one command per request.
# Each step waits for checking to finish, so the sleeps cannot overlap.
t0 = time.perf_counter()
await client.execute_command(session_id, "lemma seq_demo: True", timeout=120.0, lease_id=lease_id)
await client.execute_command(session_id, "proof -", timeout=120.0, lease_id=lease_id)
for i in range(1, N_HAVES + 1):
    await client.execute_command(session_id, slow_have(i), timeout=120.0, lease_id=lease_id)
await client.execute_command(session_id, "show True by simp", timeout=120.0, lease_id=lease_id)
await client.execute_command(session_id, "qed", timeout=120.0, lease_id=lease_id)
sequential_wall = time.perf_counter() - t0
print(f"step-by-step: {sequential_wall:.1f}s | one chunk: {parallel_wall:.1f}s "
      f"| speedup x{sequential_wall / parallel_wall:.1f}")

step-by-step: 26.6s | one chunk: 6.9s | speedup x3.8


In [9]:
await client.execute_command(session_id, "end", lease_id=lease_id)
await client.release_session(session_id, lease_id=lease_id)

{'success': True, 'session_id': '9f22572d-ef71-4c95-af4d-95c157abe754'}

To see the knobs bite, edit `.env`, restart the server, and rerun the chunk cell (the pool is rebuilt on restart, so new sessions pick up the options):

| `.env` setting | expected chunk wall time |
|---|---|
| `ISABELLE_PARALLEL_PROOFS=2`, `ISABELLE_SESSION_THREADS=4` | ≈ `SLEEP_S` (all 4 haves in flight) |
| `ISABELLE_PARALLEL_PROOFS=2`, `ISABELLE_SESSION_THREADS=2` | ≈ `2 × SLEEP_S` (two at a time) |
| `ISABELLE_PARALLEL_PROOFS=0` | ≈ `N_HAVES × SLEEP_S` (no forking — matches the step-by-step baseline) |

### Recorded results (2026-06-10)

Environment: rebuilt `isabellegym-isabelle-gym` image (Isabelle2025-2, linux/amd64 under qemu on Apple Silicon, `mem_limit: 12g`), `N_HAVES = 4`, `SLEEP_S = 2`. The first two rows are the cell outputs above; the third row reran the identical chunk after restarting the server with `ISABELLE_PARALLEL_PROOFS=0` (script: `claude-work/2026-6-10-run-parallelism-demo/run_noparallel.py`), then the server was restarted again with the default parallel settings.

| server config | run | wall time | server `execution_time` |
|---|---|---|---|
| `parallel_proofs=2`, `threads=4` | one chunk, 4 parallel `have`s | **6.9 s** | 6.66 s |
| `parallel_proofs=2`, `threads=4` | step-by-step (one command per request) | **26.6 s** | — |
| `parallel_proofs=0` (forking off) | one chunk, same 4 `have`s | **24.9 s** | 24.72 s |

Takeaways:

- With forking on, the chunk runs **×3.8** faster than the step-by-step baseline (26.6 s → 6.9 s) — the speedup matches `N_HAVES = 4`, i.e. all four `have` bodies were checked concurrently.
- With `parallel_proofs=0` the *same chunk* degrades to 24.9 s ≈ `N_HAVES × SLEEP_S` + overhead — indistinguishable from the step-by-step baseline, confirming the speedup comes from PIDE proof forking and not from chunked submission itself.
- The parallel chunk does not hit the theoretical ~`SLEEP_S` floor (2 s) because the amd64 image runs under qemu emulation on this host, which inflates per-step checking overhead (visible also in the 26.6 s baseline vs ~8 s of pure sleep).

## Big-chunk verification with `verify_chunk`

The sledgehammer section above advanced the `amc12b_2020_p22` proof **one command per call** — many round-trips, and any single line could time out on its own with no indication of *which* line hung.

`verify_chunk` submits the **whole proof body in one call**. It is checked with per-session parallelism on, bounded by a **single** wall budget, and returns a **per-command status report** in source order — each command is `ok`, `failed`, `running`, or `unprocessed`. On a timeout the report is *partial* and names the still-`running` line (the likely loop) instead of surfacing an opaque timeout.

Below we verify the **exact same proof** from the sledgehammer section as a single chunk.

In [ ]:
# verify_chunk needs the theory already begun, so acquire a session, enter the
# theory, and process the header as a normal step first.
chunk_payload = await client.acquire_session(["Complex_Main"], "HOL")
chunk_sid = chunk_payload["session_id"]
chunk_lease = chunk_payload["lease_id"]
await client.enter_theory(chunk_sid, "amc12b_2020_p22", lease_id=chunk_lease)
await client.execute_command(
    chunk_sid,
    'theory amc12b_2020_p22 imports Complex_Main "HOL-Library.Code_Target_Numeral" begin',
    lease_id=chunk_lease,
)

In [ ]:
# The same theorem..qed proof from the sledgehammer section, as ONE chunk.
proof_chunk = r"""theorem amc12b_2020_p22: fixes t :: real shows "((2 powr t - 3 * t) * t) / (4 powr t) ≤ 1 / 12"
proof -
have hpow: "(2 powr t)^2 = 4 powr t"
by (metis powr_mult numeral_Bit0_eq_double power2_eq_square)
have pos: "(0::real) < 4 powr t" by simp
have amgm: "2 * (2 powr t) * (6 * t) ≤ (2 powr t)^2 + (6 * t)^2"
by (rule sum_squares_bound)
have rearr: "12 * ((2 powr t - 3 * t) * t) = 2 * (2 powr t) * (6 * t) - (6 * t)^2"
by (simp add: power2_eq_square algebra_simps)
from amgm rearr hpow have key: "12 * ((2 powr t - 3 * t) * t) ≤ 4 powr t"
by linarith
from key pos show ?thesis
by (simp add: field_simps)
qed
"""

# Single call, single wall budget; per-command status comes back in source order.
report = await client.verify_chunk(chunk_sid, proof_chunk, timeout=180.0, lease_id=chunk_lease)

# Built-in pretty printer in the async client — no custom loop needed.
client.print_chunk_report(report)
report

In [ ]:
# Finish the theory and return the session to the pool.
await client.execute_command(chunk_sid, 'end', timeout=120.0, lease_id=chunk_lease)
await client.close_session(chunk_sid, lease_id=chunk_lease)

## Diagnostic queries with `diagnostic`

`verify_chunk` reports only whether each command is `ok`/`failed` (plus any error/warning text) — it **drops the `writeln` output** that *diagnostic* commands produce. So asking `verify_chunk` to run `thm conjI` tells you the command succeeded but **not what the theorem says**.

The `diagnostic` endpoint fills that gap. It runs a single **read-only** Isabelle query — `thm`, `term`, `prop`, `typ`, `prf`, `find_theorems`, `find_consts`, any `print_*`/`find_*` — **transiently** (insert → capture output → discard), so it returns the query's output **without changing the proof script or the rollback chain**. Input is gatekept by an allowlist of diagnostic keywords plus a denylist of code-executing / IO commands (`ML`, `setup`, `*_file`, …), which are rejected with HTTP 422.

It is exposed as `client.diagnostic(session_id, command, lease_id=...)` and, for agents, as the MCP `diagnostic` tool. The session must already have a theory begun (via `enter_theory`).

In [5]:
await client.health()

{'service': 'IsabelleGym Server',
 'version': '0.0.2',
 'status': 'healthy',
 'gateway_alive': True,
 'active_sessions': 1,
 'busy_sessions': 0,
 'max_pool_size': 2,
 'max_concurrent_sledgehammer': 4,
 'memory_management_enabled': True,
 'memory_used_mb': 1827.2,
 'memory_limit_mb': 12288.0,
 'memory_pressure_pct': 14.9,
 'timestamp': '2026-06-25T02:04:01.739756'}

In [6]:
diag_payload = await client.acquire_session(["Complex_Main"], "HOL")
diag_sid = diag_payload["session_id"]
diag_lease = diag_payload["lease_id"]
await client.enter_theory(diag_sid, "Diag_Demo", imports=["Complex_Main"], lease_id=diag_lease)
diag_payload

{'session_id': 'b1012ddd-d3dc-4ad5-a478-2e38985bf860',
 'created_at': 1782353049.3944097,
 'theories': ['Complex_Main'],
 'status': 'active',
 'reused': False,
 'lease_id': '02afb0e54e00'}

In [7]:
# Read-only queries — each returns the command's writeln output; nothing is added to the proof.
print("# thm conjI\n" + (await client.diagnostic(diag_sid, "thm conjI", lease_id=diag_lease))["output"])
print("\n# term\n" + (await client.diagnostic(diag_sid, 'term "map f (xs @ ys)"', lease_id=diag_lease))["output"])
print("\n# typ\n" + (await client.diagnostic(diag_sid, 'typ "nat ⇒ bool"', lease_id=diag_lease))["output"])
print("\n# find_theorems\n" + (await client.diagnostic(diag_sid, 'find_theorems "_ + 0 = _"', lease_id=diag_lease))["output"])
print("\n# print_statement conjI\n" + (await client.diagnostic(diag_sid, "print_statement conjI", lease_id=diag_lease))["output"])

# thm conjI
?P ⟹ ?Q ⟹ ?P ∧ ?Q

# term
"map f (xs @ ys)"
  :: "'b list"

# typ
"nat ⇒ bool"

# find_theorems
find_theorems
  "_ + 0 = _"

found 10 theorem(s):
  Groups.monoid_add_class.add_0_right: ?a + 0 = ?a
  Semiring_Normalization.comm_semiring_1_class.semiring_normalization_rules(6):
    ?a + 0 = ?a
  Groups.comm_monoid_add_class.add.comm_neutral: ?a + 0 = ?a
  SMT.verit_sum_simplify: ?a + 0 = ?a
  Int.plus_int_code(1): ?k + 0 = ?k
  SMT.z3_rule(116): ?x + 0 = ?x
  Nat.add_0_right: ?m + 0 = ?m
  SMT.z3_rule(115): ?x + 0 = ?x
  String.add_literal_code(2): ?s + STR '''' = ?s
  Code_Numeral.plus_integer_code(1): ?k + 0 = ?k

# print_statement conjI
theorem conjI:
  fixes P :: "bool"
    and Q :: "bool"
  assumes "P"
    and "Q"
  shows "P ∧ Q"


### Contrast with `verify_chunk`, transience, and the safety gate

Running the *same* `thm conjI` through `verify_chunk` confirms the command succeeds but returns **no message text** — the writeln output is not part of the proof-status report. That is exactly why `diagnostic` exists.

`diagnostic` is also **transient** (the theory source is unchanged after running queries) and **gatekept** (a code-executing or state-mutating command is rejected with HTTP 422 before it ever reaches the prover).

In [8]:
# The SAME thm conjI via verify_chunk: it "succeeds" but the report carries no writeln output —
# this is the gap the diagnostic endpoint closes.
vc = await client.verify_chunk(diag_sid, "thm conjI", lease_id=diag_lease)
print("commands:", [(c["kind"], c["status"], c["messages"]) for c in vc["commands"]])
vc

commands: [('thm', 'ok', [])]


{'success': True,
 'proof_open': False,
 'used_sorry': False,
 'timed_out': False,
 'stuck_line': None,
 'commands': [{'index': 2,
   'line': 1,
   'node_line': 2,
   'kind': 'thm',
   'status': 'ok',
   'messages': []}],
 'execution_time': 0.19413352012634277}

In [9]:
import httpx

# Transience: queries do not change the theory source the prover sees.
before = (await client.get_source(diag_sid, lease_id=diag_lease))["source"]
await client.diagnostic(diag_sid, "thm refl", lease_id=diag_lease)
await client.diagnostic(diag_sid, "print_theorems", lease_id=diag_lease)
after = (await client.get_source(diag_sid, lease_id=diag_lease))["source"]
print("source unchanged by diagnostics:", before == after)

# Safety gate: code-executing / state-mutating input is rejected (HTTP 422) before the prover.
for bad in ['ML "writeln 1"', 'lemma sneaky: "True"', "thm conjI\nML bar"]:
    try:
        await client.diagnostic(diag_sid, bad, lease_id=diag_lease)
        print("ACCEPTED (unexpected):", repr(bad))
    except httpx.HTTPStatusError as e:
        d = e.response.json().get("detail")
        msg = d[0]["msg"] if isinstance(d, list) and d else d
        print(f"[{e.response.status_code}] {repr(bad):<26} -> {msg}")

source unchanged by diagnostics: True
[422] 'ML "writeln 1"'           -> Value error, 'ML' is not an allowed diagnostic command (expected thm/term/prop/typ/prf, or a print_*/find_* inspector)
[422] 'lemma sneaky: "True"'     -> Value error, 'lemma' is not an allowed diagnostic command (expected thm/term/prop/typ/prf, or a print_*/find_* inspector)
[422] 'thm conjI\nML bar'        -> Value error, command contains a disallowed keyword 'ML' (code-execution / IO commands are not permitted)


In [10]:
await client.close_session(diag_sid, lease_id=diag_lease)

{'success': True}

In [13]:
data = await client.verify_bigstep_file(file_path="../MCP-comparison/runs/isabellegym/restrictive-prompt/mathd_algebra_276_rep4.thy", field="HOL-Computational_Algebra")

In [11]:
data.raise_for_status()
resp = data.json()
bool(resp.get("theory_verified", False))

True

In [12]:
resp.get("error", "bigstep verification failed")

## LSP-style read-only mode (file sync, per-line diagnostics, position queries)

New server surface for editor/agent-style interaction (design: `claude-work/2026-8-8-research-lsp-readonly-mode/`). The idea: **the agent edits a `.thy` file on disk; the session mirrors it** — `load_document` replaces the session's whole document with the file text (disk is the source of truth), and every query is read-only.

Features shown below:
- `load_document(..., report=True)` — whole-file load with a **per-command status report with line+column ranges** (failures are *not* rolled back, so broken state stays inspectable, LSP-style).
- `goals_at_line` / `command_at_line` — jEdit-style position queries, snapshot-based (zero document mutation).
- `get_local_facts` / `get_global_facts`, `get_last_report` — retained diagnostics.
- Queries on **finished theories** (trailing `end`) — no more probe hangs.

In [22]:
# A demo .thy file ON DISK with a deliberate error on line 4.
demo_file = "lsp_demo.thy"
with open(demo_file, "w") as f:
    f.write("""theory LspDemo imports Main begin

lemma ok1: True by simp
lemma bad: False by auto
lemma ok2: True by simp
""")

# Create a session and mirror the file into it (full-file mode: the text carries its own header).
lsp = await client.create_session(["Main"], "HOL", label="demo/lsp_demo.thy")
lsp_sid, lsp_lease = lsp["session_id"], lsp["lease_id"]

with open(demo_file) as f:
    text = f.read()
resp = await client.load_document(lsp_sid, text, timeout=120, report=True, lease_id=lsp_lease)
print("load success:", resp["success"], "| theory:", resp["theory"], "| error:", resp["error"])
print()
print(client.format_chunk_report(resp["report"]))

load success: False | theory: LspDemo | error: line 4: Failed to apply initial proof method⌂:
goal (1 subgoal):
 1. False

verify_chunk: success=False proof_open=False used_sorry=False timed_out=False stuck_line=None time=0.00s
  [OK ] line   1  theory  ok
  [OK ] line   3  lemma   ok
  [OK ] line   3  by      ok
  [OK ] line   4  lemma   ok
  [ERR] line   4  by      failed
        error: Failed to apply initial proof method⌂: goal (1 subgoal): 1. False
  [OK ] line   5  lemma   ok
  [OK ] line   5  by      ok


In [23]:
# The report carries a per-command RANGE (1-based line/col, UTF-16 columns) —
# the failing command's span maps directly onto the editor buffer.
for c in resp["report"]["commands"]:
    if c["status"] == "failed":
        r = c["range"]
        print(f'FAILED: kind={c["kind"]} at lines {r["start"]["line"]}:{r["start"]["col"]}–{r["end"]["line"]}:{r["end"]["col"]}')
        print("message:", c["messages"][0]["text"][:120])
# Note: failures are NOT rolled back — ok2 (after the error) still checked, and ok1 is a live fact:
(await client.diagnostic(lsp_sid, "thm ok1", lease_id=lsp_lease))["output"]

FAILED: kind=by at lines 4:18–4:25
message: Failed to apply initial proof method⌂:
goal (1 subgoal):
 1. False


'True'

In [24]:
# The file-sync model: edit the file ON DISK, reload — the session follows the file.
with open(demo_file, "w") as f:
    f.write("""theory LspDemo imports Main begin

lemma ok1: True by simp
lemma bad: False by simp  (* still wrong, but now it fails differently *)
""")

with open(demo_file) as f:
    text2 = f.read()
resp2 = await client.load_document(lsp_sid, text2, timeout=120, report=True, lease_id=lsp_lease)
print("reload success:", resp2["success"], "| error:", resp2["error"])

# The last report is retained server-side — queryable later without re-running anything:
last = await client.get_last_report(lsp_sid, lease_id=lsp_lease)
print("last_report commands:", [(c["kind"], c["status"]) for c in last["report"]["commands"]])

reload success: False | error: line 4: Failed to apply initial proof method⌂:
goal (1 subgoal):
 1. False
last_report commands: [('theory', 'ok'), ('lemma', 'ok'), ('by', 'ok'), ('lemma', 'ok'), ('by', 'failed')]


### Position queries (snapshot-based, read-only)

`goals_at_line` returns the proof state **before/after** the command containing a line (like `lean_goal`); `command_at_line` returns the enclosing command with its range. Both read retained per-command results from an immutable snapshot — **no probes, no document mutation**, and they work past a trailing `end`.

In [25]:
proof_text = """theory GoalDemo imports Main begin
lemma conj_swap: "A ∧ B ⟹ B ∧ A"
proof (rule conjI)
  assume "A ∧ B"
  show B by (rule conjunct2[OF ‹A ∧ B›])
  show A by (rule conjunct1[OF ‹A ∧ B›])
qed
"""
await client.load_document(lsp_sid, proof_text, timeout=120, lease_id=lsp_lease)

for line in [2, 3, 5]:
    g = await client.goals_at_line(lsp_sid, line, lease_id=lsp_lease)
    print(f"--- line {line}: {g['command']['kind']!r:12} range={g['command']['range']['start']['line']}")
    print("  before:", (g["goals_before"] or ["(none)"])[0][:80].replace("\n", " ⏎ "))
    print("  after: ", (g["goals_after"] or ["(none)"])[0][:80].replace("\n", " ⏎ "))

# A comment/blank line resolves to the enclosing command (ignored-span aware):
cmd = await client.command_at_line(lsp_sid, 4, lease_id=lsp_lease)
print("line 4 (assume):", cmd["kind"], "->", cmd["source"][:50])

--- line 2: 'lemma'      range=2
  before: (none)
  after:  proof (prove) ⏎ goal (1 subgoal): ⏎  1. A ∧ B ⟹ B ∧ A
--- line 3: 'proof'      range=3
  before: proof (prove) ⏎ goal (1 subgoal): ⏎  1. A ∧ B ⟹ B ∧ A
  after:  proof (state) ⏎ goal (2 subgoals): ⏎  1. A ∧ B ⟹ B ⏎  2. A ∧ B ⟹ A
--- line 5: 'assume'     range=4
  before: proof (state) ⏎ goal (2 subgoals): ⏎  1. A ∧ B ⟹ B ⏎  2. A ∧ B ⟹ A
  after:  proof (state) ⏎ this: ⏎   A ∧ B ⏎  ⏎ goal (2 subgoals): ⏎  1. A ∧ B ⟹ B ⏎  2. A ∧ B ⟹ A
line 4 (assume): proof -> proof (rule conjI)


In [26]:
# Fact queries mid-proof (transient probes): local context vs theory-level facts.
await client.load_document(
    lsp_sid,
    'theory FactDemo imports Main begin\nlemma "⟦ A; B ⟧ ⟹ A ∧ B"\nproof -\n  assume A B',
    timeout=120, lease_id=lsp_lease,
)
print("local facts:", (await client.get_local_facts(lsp_sid, lease_id=lsp_lease))["facts"])
print()
glob = await client.get_global_facts(lsp_sid, limit=5, lease_id=lsp_lease)
print(f"global facts (first {glob['count']} of Main):")
for f in glob["facts"]:
    print("  ", f[:70])

local facts: ['??.<unnamed>:\nA\nB', 'this:\nA\nB']

global facts (first 5 of Main):
   exE: \<exists>x. ?P x \<Longrightarrow> (\<And>x. ?P x \<Longrightarro
   exI: ?P ?x \<Longrightarrow> \<exists>x. ?P x
   ext: (\<And>x. ?f x = ?g x) \<Longrightarrow> ?f = ?g
   mp: ?P \<longrightarrow> ?Q \<Longrightarrow> ?P \<Longrightarrow> ?Q
   sym: ?s = ?t \<Longrightarrow> ?t = ?s


In [27]:
# Finished theories: after a trailing `end`, queries used to hang (2× channel timeout → 500).
# Now they answer immediately — probes insert BEFORE the `end`; subgoals/in_proof short-circuit.
finished = """theory DoneDemo imports Main begin
lemma foo: True by simp
end
"""
await client.load_document(lsp_sid, finished, timeout=120, report=True, lease_id=lsp_lease)

import time
t0 = time.time()
out = (await client.diagnostic(lsp_sid, "thm foo", lease_id=lsp_lease))["output"]
subs = await client.get_subgoals(lsp_sid, lease_id=lsp_lease)
print(f"thm foo -> {out!r}  ({time.time()-t0:.2f}s)")
print("subgoals after end:", subs["subgoals"], "| proof_finished:", subs["proof_finished"])

thm foo -> 'True'  (0.25s)
subgoals after end: [] | proof_finished: True


## Heap pool: verified, shared imports with task-group tenancy

Imports are **static, verified heaps** (design: `docs/DESIGN_CHOICES.md` §1.13): each project is built once with `isabelle build -b` — **build passing IS the verification gate** — and every session in the owning **task group** shares the heap image. Agents write qualified imports (`imports "DemoHeap.Shared"`); fact names stay unqualified. Source changes mark the heap **stale** (422) until an explicit rebuild.

Below: a mini project written into this repo (visible in the container at `/app/demo/...`), built as a heap, imported from a fresh session.

In [28]:
import os
# The repo is volume-mounted at /app in the container, so files written here are visible to the server.
# Anchor to the repo root (found by the setup cell) so this works no matter where
# Jupyter was launched; the repo is volume-mounted at /app in the container.
proj_host = os.path.join(REPO_ROOT, "demo", "heap_demo_project")
proj_container = "/app/demo/heap_demo_project"
os.makedirs(proj_host, exist_ok=True)
with open(f"{proj_host}/Shared.thy", "w") as f:
    f.write("""theory Shared imports Main begin
lemma shared_lemma: True by simp
end
""")

# Build the verified heap for task group "demo" (409 while building; failure -> status=failed with log tail).
build = await client.heap_build("demo", proj_container, session_name="DemoHeap")
print("heap status:", build["status"], "| session:", build["session_name"])

heap status: ready | session: DemoHeap


In [29]:
# The manifest is the inspection record: which sources, which hashes, which ROOT.
man = await client.get_heap("demo", proj_container)
print("fingerprint:", man["fingerprint"][:16], "| status:", man["status"])
print("ROOT:\n" + man["root_text"])
for tf in man["theory_files"]:
    print(f'  {tf["path"]}  sha256:{tf["sha256"][:12]}…')

fingerprint: e151e38c033efec3 | status: ready
ROOT:
session DemoHeap = HOL +
  directories "/app/demo/heap_demo_project"
  theories
    Shared

  /app/demo/heap_demo_project/Shared.thy  sha256:3f7b422a23bc…


In [30]:
# Sessions in group "demo" start ON the heap and import it with qualified names.
heap_sess = await client.create_session(task_group="demo", heap_session="DemoHeap")
h_sid, h_lease = heap_sess["session_id"], heap_sess["lease_id"]
print("session theories:", heap_sess["theories"], "| group:", heap_sess["task_group"])

doc = """theory Consumer imports "DemoHeap.Shared" begin
lemma use_it: True by (simp add: shared_lemma)
"""
r = await client.load_document(h_sid, doc, timeout=120, report=True, lease_id=h_lease)
print("load success:", r["success"])
print("thm shared_lemma ->", (await client.diagnostic(h_sid, "thm shared_lemma", lease_id=h_lease))["output"])

session theories: ['DemoHeap.Shared'] | group: demo
load success: True
thm shared_lemma -> True


In [31]:
# Tenancy: a session in ANOTHER task group cannot use the demo heap (403).
import httpx
try:
    await client.create_session(task_group="other", heap_session="DemoHeap")
except httpx.HTTPStatusError as e:
    print("cross-group refused:", e.response.status_code, "-", e.response.json()["detail"][:90])

# Staleness: edit the source — the heap flips to stale and session creation is refused (422).
with open(f"{proj_host}/Shared.thy", "w") as f:
    f.write("""theory Shared imports Main begin
lemma shared_lemma: True by simp
lemma shared_lemma2: True by simp
end
""")
try:
    await client.create_session(task_group="demo", heap_session="DemoHeap")
except httpx.HTTPStatusError as e:
    print("stale refused:", e.response.status_code, "-", e.response.json()["detail"][:90])

# Explicit rebuild (keeps the session name) — then the new fact is available.
rebuild = await client.heap_build("demo", proj_container)
print("rebuild:", rebuild["status"], "| session:", rebuild["session_name"])
heap_sess2 = await client.create_session(task_group="demo", heap_session="DemoHeap")
h2_sid, h2_lease = heap_sess2["session_id"], heap_sess2["lease_id"]
await client.load_document(h2_sid, doc, timeout=120, lease_id=h2_lease)
print("thm shared_lemma2 ->", (await client.diagnostic(h2_sid, "thm shared_lemma2", lease_id=h2_lease))["output"])

cross-group refused: 403 - heap DemoHeap belongs to another task group; sessions may only use their own group's heaps
stale refused: 422 - heap DemoHeap is stale (sources changed since build); rebuild with POST /api/v1/heaps/buil
rebuild: ready | session: DemoHeap
thm shared_lemma2 -> True


In [32]:
# Admin surface: group listing, and the static console at GET /admin (open in a browser).
print(await client.list_heap_groups())

# Cleanup: close sessions and delete the demo heap (GC's the on-disk image).
await client.close_session(lsp_sid, lease_id=lsp_lease)
await client.close_session(h_sid, lease_id=h_lease)
await client.close_session(h2_sid, lease_id=h2_lease)
print(await client.delete_heap_group("demo"))

{'groups': [{'task_group': 'demo', 'heap_count': 1, 'ready': 1}, {'task_group': 'mcp', 'heap_count': 1, 'ready': 1}]}
{'deleted': 1, 'task_group': 'demo'}


## Hover, go-to-definition & positioned sledgehammer

Three more LSP-style queries (all read-only, all position-explicit — 1-based line/col, UTF-16 columns):

- `hover_at(line, col)` — type/info of the symbol at a position (Pure `Rendering`, snapshot-based, no evaluation).
- `definition_at(line, col)` — where a symbol is defined: file targets into library/project sources (`~~/` expanded), or in-node positions for the document's own entities.
- `sledgehammer_at(line, subgoal)` — sledgehammer on the goal at a **specific line and subgoal**, via an overlay print function (no text insertion, no tip requirement).

In [6]:
import client.async_client as ac
print(ac.__file__)
print(hasattr(ac.IsabelleGymAsyncClient, "hover_at"))

/Users/winstonren/GitHub/IsabelleGym/client/async_client.py
True


In [7]:
hv = await client.create_session(["Main"], "HOL", label="demo/hover")
hv_sid, hv_lease = hv["session_id"], hv["lease_id"]

hv_text = """theory HoverDemo imports Main begin
lemma hd_id: "hd [x] = x" by simp
lemma use_hd: "hd [1::nat] = 1" by (simp add: hd_id)
"""
await client.load_document(hv_sid, hv_text, timeout=120, lease_id=hv_lease)

# hover over `hd` on line 3 (compute the column from the text — no magic numbers)
line_no = 3
line_text = hv_text.splitlines()[line_no - 1]
col = line_text.index('"hd') + 2  # 1-based col of the constant hd (first "hd" occurrence is inside use_hd)
h = await client.hover_at(hv_sid, line_no, col, lease_id=hv_lease)
print(f"hover @{line_no}:{col} ->")
for c in h["contents"]:
    print("  ", c)

# hover over our OWN lemma reference `hd_id`
col2 = line_text.index("hd_id") + 1
h2 = await client.hover_at(hv_sid, line_no, col2, lease_id=hv_lease)
print(f"hover @{line_no}:{col2} ->", h2["contents"])

hover @3:16 ->
   constant "List.list.hd"
   :: nat list ⇒ nat
hover @3:47 -> ['fact "HoverDemo.hd_id"']


In [8]:
# go-to-definition: library constant -> file position in the HOL sources;
# our own lemma -> in-node position in the current document.
d1 = await client.definition_at(hv_sid, line_no, col, lease_id=hv_lease)
print("definition of hd:")
for t in d1["targets"]:
    if t["kind"] == "file":
        print(f'  file: {t["file"]}  line {t["line"]}')
    else:
        print(f'  node: {t["node"]}  lines {t["start_line"]}-{t["end_line"]}')

d2 = await client.definition_at(hv_sid, line_no, col2, lease_id=hv_lease)
print("definition of hd_id:", d2["targets"])

definition of hd:
  file: /opt/isabelle/src/HOL/List.thy  line 13
definition of hd_id: [{'kind': 'node', 'file': None, 'line': None, 'col': None, 'end_line': 2, 'end_col': None, 'node': 'Draft.HoverDemo', 'start_line': 2, 'id': None}]


In [9]:
# sledgehammer at a SPECIFIC line + subgoal (overlay op; the document is untouched).
sh_text = """theory HammerDemo imports Main begin
lemma two_goals: "⟦ A; B ⟧ ⟹ A ∧ B"
proof (rule conjI)
"""
await client.load_document(hv_sid, sh_text, timeout=120, lease_id=hv_lease)

# line 3 is `proof (rule conjI)` — after it, two subgoals are open.
for sg in (1, 2):
    r = await client.sledgehammer_at(hv_sid, 3, subgoal=sg, timeout_s=20, lease_id=hv_lease)
    print(f"subgoal {sg}: found={r['found']}")
    for s in (r.get("results") or [])[:3]:
        print("   ", s[:90])

# a non-proof line answers cleanly (no timeout, no hang):
r_none = await client.sledgehammer_at(hv_sid, 1, lease_id=hv_lease)
print("line 1:", r_none["found"], "-", r_none.get("error"))

subgoal 1: found=True
    linarith found a proof...satx found a proof...satx: Try this: apply satx (2 ms)presburger 
subgoal 2: found=True
    satx found a proof...simp found a proof...simp: Try this: apply simp (1 ms)argo: Try this:
line 1: False - no open goal at line 1


In [10]:
# regression sanity: the tip sledgehammer (existing tool) is unchanged.
# (tip mode needs a plain open goal — load a bare lemma statement for it)
await client.load_document(hv_sid, "theory TipDemo imports Main begin\nlemma tip_goal: True", timeout=120, lease_id=hv_lease)
tip = await client.sledgehammer(hv_sid, timeout_s=20, lease_id=hv_lease)
print("tip sledgehammer:", tip.get("success"), (tip.get("suggestions") or tip.get("results") or [])[:2])

await client.close_session(hv_sid, lease_id=hv_lease)
print("session closed")

tip sledgehammer: True ['satx found a proof...', 'satx: Try this: by satx (0.4 ms)']
session closed


## LSP-style MCP server (`mcp_lsp_server`) — full tool tour

Everything above drove the HTTP API directly. This section talks to the **MCP server** the way an LLM agent harness does: it spawns `python -m mcp_lsp_server.app` as a stdio subprocess and calls its tools over the MCP protocol — no HTTP client involved.

The workflow is the lean-lsp-mcp file-sync model: tools take a `file_path` (never a session id), the MCP re-reads the file from disk before every call and re-syncs the session (`load_document(report=true)`), and execution tools (`isabelle_multi_attempt`, `isabelle_run_code`) run on a warm **scratch** pool — never on the file's session. The MCP never writes files; the notebook writes the demo file itself.

Prerequisites: the gym HTTP server on `:8000` (as for all sections above) and `pip install "mcp<2"` in the notebook's environment (mcp 2.0 removed `mcp.server.fastmcp`).


In [56]:
import os, sys, json
from contextlib import AsyncExitStack
from mcp import ClientSession
from mcp.client.stdio import StdioServerParameters, stdio_client

MCP_DEMO = os.path.join(REPO_ROOT, "demo", "mcp_lsp_demo.thy")

# v1: a deliberate error on line 4, for the diagnostics demo.
V1 = 'theory McpLspDemo imports Main begin\n\nlemma ok1: True by simp\nlemma broken: False by simp\n\nend\n'
with open(MCP_DEMO, "w") as f:
    f.write(V1)

async def mcp_call(name, **args):
    """Call an MCP tool and print its text payload (tools return JSON strings)."""
    result = await mcp_sess.call_tool(name, args)
    text = "\n".join(c.text for c in result.content if getattr(c, "type", "") == "text")
    print(f"--- {name} ---")
    print(text[:1500])
    print()
    return text


In [57]:
# Spawn the MCP server (host-side subprocess) and connect over stdio.
# PYTHONPATH must include the repo root so `mcp_lsp_server` and `client` import.
stack = AsyncExitStack()
env = dict(os.environ, PYTHONPATH=REPO_ROOT + os.pathsep + os.environ.get("PYTHONPATH", ""))
params = StdioServerParameters(command=sys.executable, args=["-m", "mcp_lsp_server.app"], env=env)
read, write = await stack.enter_async_context(stdio_client(params))
mcp_sess = await stack.enter_async_context(ClientSession(read, write))
await mcp_sess.initialize()

tools = await mcp_sess.list_tools()
print(f"{len(tools.tools)} tools:")
print(", ".join(sorted(t.name for t in tools.tools)))


23 tools:
isabelle_build_heap, isabelle_checkpoint, isabelle_close, isabelle_command_at_line, isabelle_definition, isabelle_diagnostic_messages, isabelle_global_facts, isabelle_goal, isabelle_heap_status, isabelle_history, isabelle_hover_info, isabelle_last_report, isabelle_local_facts, isabelle_multi_attempt, isabelle_open, isabelle_proof_state, isabelle_query, isabelle_restore, isabelle_rollback, isabelle_run_code, isabelle_sledgehammer, isabelle_source, isabelle_sync


In [58]:
# Open the file (binds a session + first sync) and read the diagnostics:
# the broken lemma shows up with its 1-based line/col range (UTF-16 columns).
await mcp_call("isabelle_open", file_path=MCP_DEMO)
await mcp_call("isabelle_diagnostic_messages", file_path=MCP_DEMO)


--- isabelle_open ---
{"opened": "/Users/winstonren/GitHub/IsabelleGym/demo/mcp_lsp_demo.thy", "task_group": "default", "heap_session": null, "synced": true, "success": false, "proof_open": false, "commands": 6, "error": null}

--- isabelle_diagnostic_messages ---
{"messages": [{"line": 4, "node_line": 4, "col": 21, "end_line": 4, "end_col": 28, "kind": "by", "status": "failed", "severity": "error", "text": "Failed to apply initial proof method\u2302:\ngoal (1 subgoal):\n 1. False"}], "success": false, "proof_open": false, "used_sorry": false}



'{"messages": [{"line": 4, "node_line": 4, "col": 21, "end_line": 4, "end_col": 28, "kind": "by", "status": "failed", "severity": "error", "text": "Failed to apply initial proof method\\u2302:\\ngoal (1 subgoal):\\n 1. False"}], "success": false, "proof_open": false, "used_sorry": false}'

In [59]:
# The file-sync model: fix the file ON DISK, then sync — the session follows the file.
V2 = 'theory McpLspDemo imports Main begin\n\nlemma hd_id: "hd [x] = x" by simp\nlemma use_hd: "hd [1::nat] = 1"\n  by (simp add: hd_id)\n\nlemma two_goals: "\\<lbrakk> A; B \\<rbrakk> \\<Longrightarrow> A \\<and> B"\nproof (rule conjI)\n  show A by assumption\n  show B by assumption\nqed\n\nend\n'
with open(MCP_DEMO, "w") as f:
    f.write(V2)
await mcp_call("isabelle_sync", file_path=MCP_DEMO)
await mcp_call("isabelle_diagnostic_messages", file_path=MCP_DEMO)  # clean now


--- isabelle_sync ---
{"synced": true, "success": false, "proof_open": false, "error": null, "stuck_line": false}

--- isabelle_diagnostic_messages ---
{"messages": [{"line": 9, "node_line": 9, "col": 10, "end_line": 9, "end_col": 23, "kind": "by", "status": "failed", "severity": "error", "text": "Failed to apply initial proof method\u2302:\ngoal (1 subgoal):\n 1. A"}, {"line": 10, "node_line": 10, "col": 10, "end_line": 10, "end_col": 23, "kind": "by", "status": "failed", "severity": "error", "text": "Failed to apply initial proof method\u2302:\ngoal (1 subgoal):\n 1. B"}], "success": false, "proof_open": false, "used_sorry": false}



'{"messages": [{"line": 9, "node_line": 9, "col": 10, "end_line": 9, "end_col": 23, "kind": "by", "status": "failed", "severity": "error", "text": "Failed to apply initial proof method\\u2302:\\ngoal (1 subgoal):\\n 1. A"}, {"line": 10, "node_line": 10, "col": 10, "end_line": 10, "end_col": 23, "kind": "by", "status": "failed", "severity": "error", "text": "Failed to apply initial proof method\\u2302:\\ngoal (1 subgoal):\\n 1. B"}], "success": false, "proof_open": false, "used_sorry": false}'

In [60]:
# Position queries (snapshot-based, read-only; 1-based lines).
# Line 8 is `proof (rule conjI)` — after it, two subgoals are open.
await mcp_call("isabelle_goal", file_path=MCP_DEMO, line=8)
await mcp_call("isabelle_command_at_line", file_path=MCP_DEMO, line=8)
await mcp_call("isabelle_proof_state", file_path=MCP_DEMO)   # tip state (no open proof)
src = await mcp_call("isabelle_source", file_path=MCP_DEMO)  # source as the session sees it


--- isabelle_goal ---
{"found": true, "command": {"kind": "proof", "source": "proof (rule conjI)", "range": {"start": {"line": 8, "col": 1}, "end": {"line": 8, "col": 19}}}, "goals_before": ["proof (prove)\ngoal (1 subgoal):\n 1. A \\<Longrightarrow> B \\<Longrightarrow> A \\<and> B"], "goals_after": ["proof (state)\ngoal (2 subgoals):\n 1. A \\<Longrightarrow> B \\<Longrightarrow> A\n 2. A \\<Longrightarrow> B \\<Longrightarrow> B"], "error": null}

--- isabelle_command_at_line ---
{"found": true, "kind": "proof", "source": "proof (rule conjI)", "range": {"start": {"line": 8, "col": 1}, "end": {"line": 8, "col": 19}}, "error": null}

--- isabelle_proof_state ---
{"subgoals": [], "proof_finished": true, "pending_qed": false, "current_theory": "McpLspDemo"}

--- isabelle_source ---
{"source": "theory McpLspDemo imports Main begin\n\nlemma hd_id: \"hd [x] = x\" by simp\nlemma use_hd: \"hd [1::nat] = 1\"\n  by (simp add: hd_id)\n\nlemma two_goals: \"\\<lbrakk> A; B \\<rbrakk> \\<Longright

In [61]:
# Transient read-only probes (never touch the proof script):
# thm/term/find_theorems-style queries, and local vs theory-level fact lists.
await mcp_call("isabelle_query", file_path=MCP_DEMO, command="thm hd_id")
await mcp_call("isabelle_local_facts", file_path=MCP_DEMO)
await mcp_call("isabelle_global_facts", file_path=MCP_DEMO, limit=5)


--- isabelle_query ---
{"success": true, "output": "hd [?x] = ?x", "error": null, "execution_time": 0.11554360389709473}

--- isabelle_local_facts ---
{"facts": [], "count": 0}

--- isabelle_global_facts ---
{"facts": [], "count": 0}



'{"facts": [], "count": 0}'

In [62]:
# hover / go-to-definition at a position (columns computed from the text — no magic numbers).
line_no = 3
line_text = V2.splitlines()[line_no - 1]
col_hd = line_text.index('"hd') + 2      # the library constant `hd`
col_own = line_text.index("hd_id") + 1   # our own lemma `hd_id`
await mcp_call("isabelle_hover_info", file_path=MCP_DEMO, line=line_no, column=col_hd)
await mcp_call("isabelle_definition", file_path=MCP_DEMO, line=line_no, column=col_hd)
await mcp_call("isabelle_definition", file_path=MCP_DEMO, line=line_no, column=col_own)


--- isabelle_hover_info ---
{"found": true, "range": {"start": {"line": 3, "col": 15}, "end": {"line": 3, "col": 17}}, "contents": ["constant \"List.list.hd\"", ":: 'a list \\<Rightarrow> 'a"], "error": null}

--- isabelle_definition ---
{"found": true, "targets": [{"kind": "file", "file": "/opt/isabelle/src/HOL/List.thy", "line": 13, "col": 11, "end_line": 13, "end_col": 13, "node": null, "start_line": null, "id": null}], "error": null}

--- isabelle_definition ---
{"found": false, "targets": [], "error": "no definition at line 3 col 7"}



'{"found": false, "targets": [], "error": "no definition at line 3 col 7"}'

In [63]:
# Positioned sledgehammer: the goal at line 8, one subgoal at a time (overlay op —
# the document is never modified). Slow by nature; timeout bounded per call.
await mcp_call("isabelle_sledgehammer", file_path=MCP_DEMO, line=8, subgoal=1, timeout_s=20)
await mcp_call("isabelle_sledgehammer", file_path=MCP_DEMO, line=8, subgoal=2, timeout_s=20)
# Tip mode (no line) answers cleanly even past the trailing `end`:
await mcp_call("isabelle_sledgehammer", file_path=MCP_DEMO, timeout_s=10)


--- isabelle_sledgehammer ---
{"found": true, "results": ["satx found a proof...argo found a proof...argo: Try this: apply argo (0.2 ms)presburger found a proof...presburger: Try this: apply presburger (0.2 ms)simp found a proof...simp: Try this: apply simp (0.3 ms)auto found a proof...auto: Try this: by auto (0.7 ms)satx: Try this: apply satx (0.6 ms)linarith found a proof...linarith: Try this: apply linarith (1 ms)metis found a proof...metis: Try this: apply metis (5 ms)DoneTry this: apply satx (0.6 ms)"], "error": null, "execution_time": 1.1363673210144043}

--- isabelle_sledgehammer ---
{"found": true, "results": ["presburger found a proof...presburger: Try this: apply presburger (0.9 ms)linarith found a proof...satx found a proof...satx: Try this: apply satx (1 ms)linarith: Try this: apply linarith (0.9 ms)auto found a proof...blast found a proof...blast: Try this: apply blast (0.2 ms)fast found a proof...fast: Try this: apply fast (1 ms)simp found a proof...fastforce found a proo

'{"success": false, "suggestions": [], "raw_output": "", "execution_time": 0.004291057586669922}'

In [64]:
# Checkpoints / history on the file session.
cp = await mcp_call("isabelle_checkpoint", file_path=MCP_DEMO)
cp_id = json.loads(cp)["checkpoint_id"]
await mcp_call("isabelle_history", file_path=MCP_DEMO)
await mcp_call("isabelle_last_report", file_path=MCP_DEMO)  # per-command report of the last sync
await mcp_call("isabelle_restore", file_path=MCP_DEMO, checkpoint_id=cp_id)
await mcp_call("isabelle_rollback", file_path=MCP_DEMO)     # nothing past the tip -> clean failure/success


--- isabelle_checkpoint ---
{"checkpoint_id": 0, "timestamp": 1787124519.0478766}

--- isabelle_history ---
{"session_id": "a4fc0822-9b6d-483e-89e9-92cd74977d22", "total_commands": 1, "history": [{"type": "document_load", "command": "theory McpLspDemo imports Main begin lemma hd_id: \"hd [x] = x\" by simp lemma use_hd: \"hd [1::nat] = 1\" by (simp add: ...", "timestamp": 1787124506.722309, "success": false, "subgoal_error": null, "subgoals_count": 0}]}

--- isabelle_last_report ---
{"report": {"success": false, "timed_out": false, "elapsed_ms": 204, "pending_qed": false, "commands": [{"messages": [], "line": 1, "node_line": 1, "i": 0, "kind": "theory", "range": {"start": {"line": 1, "col": 1}, "end": {"line": 1, "col": 37}}, "status": "ok"}, {"messages": [], "line": 3, "node_line": 3, "i": 2, "kind": "lemma", "range": {"start": {"line": 3, "col": 1}, "end": {"line": 3, "col": 26}}, "status": "ok"}, {"messages": [], "line": 3, "node_line": 3, "i": 4, "kind": "by", "range": {"start": {"l

'{"success": true, "output": ""}'

In [65]:
# Scratch execution — isolated warm sessions, the file session is never touched.
# multi_attempt: which proof step closes line 5's goal? (candidate-relative error lines)
await mcp_call("isabelle_multi_attempt", file_path=MCP_DEMO, line=5,
               candidates=["by simp", "by (simp add: hd_id)", "by (rule conjI)"])
# run_code: an independent snippet (body of a scratch theory importing Main).
await mcp_call("isabelle_run_code", chunk="lemma scratch_demo: True by simp")


--- isabelle_multi_attempt ---
[{"candidate": "by simp", "success": true, "proof_open": false, "pending_qed": false, "used_sorry": false, "timed_out": false, "failed": []}, {"candidate": "by (simp add: hd_id)", "success": true, "proof_open": false, "pending_qed": false, "used_sorry": false, "timed_out": false, "failed": []}, {"candidate": "by (rule conjI)", "success": false, "proof_open": false, "pending_qed": false, "used_sorry": false, "timed_out": false, "failed": [{"candidate_line": 1, "kind": "by", "status": "failed", "messages": "Failed to apply initial proof method\u2302:\ngoal (1 subgoal):\n 1. hd [1] = 1"}]}]

--- isabelle_run_code ---
{"success": true, "output": "", "error": null, "proof_open": false, "used_sorry": false, "failed": []}



'{"success": true, "output": "", "error": null, "proof_open": false, "used_sorry": false, "failed": []}'

In [66]:
# Heap tools: build a verified heap for the demo project (LONG-RUNNING — isabelle
# build -b), inspect its manifest, then open a consumer file bound to that heap.
# The project path must be visible to the SERVER — the repo is mounted at /app in
# the container. The consumer file itself is read from disk by the MCP (host path).
build = await mcp_call("isabelle_build_heap", task_group="demo-mcp",
                       project="/app/demo/heap_demo_project", session_name="McpDemoHeap")
await mcp_call("isabelle_heap_status", task_group="demo-mcp")

CONSUMER = 'theory McpHeapConsumer imports "McpDemoHeap.Shared" begin\nlemma use_it: True by (simp add: shared_lemma)\nend\n'
CONSUMER_FILE = os.path.join(REPO_ROOT, "demo", "mcp_heap_consumer.thy")
with open(CONSUMER_FILE, "w") as f:
    f.write(CONSUMER)
await mcp_call("isabelle_open", file_path=CONSUMER_FILE,
               task_group="demo-mcp", heap_session="McpDemoHeap")
await mcp_call("isabelle_query", file_path=CONSUMER_FILE, command="thm shared_lemma")


--- isabelle_build_heap ---
{"task_group": "demo-mcp", "project": "/app/demo/heap_demo_project", "session_name": "McpDemoHeap", "root_dir": "/root/.isabelle/heap_pool/_roots/demo-mcp/fb75df0ef77e254f", "fingerprint": "1d8cd79bb7fc543f7ac215609a9fbbf7b8a71d2ca15740ed874f2d267ee0d80f", "status": "ready", "built_at": 1787124545.3849654, "built_by": "demo-mcp", "build_log_tail": "0:00:03 elapsed time\n"}

--- isabelle_heap_status ---
{"heaps": [{"task_group": "demo-mcp", "project": "/app/demo/heap_demo_project", "session_name": "McpDemoHeap", "root_dir": "/root/.isabelle/heap_pool/_roots/demo-mcp/fb75df0ef77e254f", "fingerprint": "1d8cd79bb7fc543f7ac215609a9fbbf7b8a71d2ca15740ed874f2d267ee0d80f", "status": "ready", "built_at": 1787124545.3849654, "built_by": "demo-mcp", "build_log_tail": "0:00:03 elapsed time\n"}]}

--- isabelle_open ---
{"opened": "/Users/winstonren/GitHub/IsabelleGym/demo/mcp_heap_consumer.thy", "task_group": "demo-mcp", "heap_session": "McpDemoHeap", "synced": true, "su

'{"success": true, "output": "True", "error": null, "execution_time": 0.10367584228515625}'

In [67]:
# Cleanup: unbind the files (sessions go back to the server pool for warm reuse).
# The demo-mcp heap persists — delete it via GET /admin or the HTTP heap
# endpoints if you don't want it.
await mcp_call("isabelle_close", file_path=CONSUMER_FILE)
await mcp_call("isabelle_close", file_path=MCP_DEMO)
# Graceful aclose() crosses Jupyter's per-cell task boundary — anyio cancel
# scopes are task-bound, so exiting a stack entered in another cell raises
# "Attempted to exit cancel scope in a different task". The sessions are
# already released above, so fall back to terminating the MCP subprocess.
try:
    await stack.aclose()
except BaseException as e:
    print(f"graceful aclose unavailable across cells ({type(e).__name__}); terminating subprocess")
    import subprocess as sp
    for pid in sp.check_output(["pgrep", "-f", "mcp_lsp_server.app"]).split():
        os.kill(int(pid), 15)
print("MCP server shut down.")


--- isabelle_close ---
{"closed": true, "file": "/Users/winstonren/GitHub/IsabelleGym/demo/mcp_heap_consumer.thy"}

--- isabelle_close ---
{"closed": true, "file": "/Users/winstonren/GitHub/IsabelleGym/demo/mcp_lsp_demo.thy"}

graceful aclose unavailable across cells (RuntimeError); terminating subprocess


  + Exception Group Traceback (most recent call last):
  |   File "/Users/winstonren/anaconda3/envs/IsabelleGym/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 772, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/winstonren/anaconda3/envs/IsabelleGym/lib/python3.11/site-packages/mcp/client/stdio/__init__.py", line 189, in stdio_client
    |     yield read_stream, write_stream
    |   File "/Users/winstonren/anaconda3/envs/IsabelleGym/lib/python3.11/contextlib.py", line 728, in __aexit__
    |     cb_suppress = await cb(*exc_details)
    |                   ^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/winstonren/anaconda3/envs/IsabelleGym/lib/python3.11/site-packages/mcp/shared/session.py", line 238, in __aexit__
    |     return await self._task_group.__aexit__(exc_type, exc_val, exc_tb)
    |         